### PR TITLE
Better checkpoints

### PR DESCRIPTION
**Pull request by jlamypoirier**
* Add a FieldUpdate tool to simplify field overrides.
* Standardize checkpoint saving and loading configs, using mixins so training checkpoint/export configs can use just the right fields (ex. no path since it's provided by the trainer)
* Adjust some fields so they make sense for both saving and loading, and merge the ones for config selection into an enum (see below).
* Move checkpoint save/load from Run to Trainer and simplify.
* Make IntervalConfig into a proper class (use FieldUpdate foe doc instead)
* Make export checkpoint fully configurable, save separately instead of using a symlink.
* Add keep_every for checkpoints to get the old behaviour, and optional callback script.

RENAME[BW COMPATIBLE] pretrained.imported_type ->pretrained.model_type
RENAME[BW COMPATIBLE] pretrained.load_weights -> pretrained.model_weights
RENAME[BW COMPATIBLE] pretrained.load_optimizer -> pretrained.optimizer_state
MERGE (pretrained.override_architecture, pretrained.load_full_base_model_config, pretrained.load_full_fast_llm_config) ->pretrained.load_config:enum

### CODE REVIEW

File: fast_llm/engine/config_utils/checkpoint.py [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+# TODO: Use packaging.version? (Safer but extra requirement)
+CHECKPOINT_VERSION = "0.1"
+KNOWN_CHECKPOINT_VERSIONS = ("0", "0.1")
+
+
+class CheckpointFormat(str, enum.Enum):
+    # Distributed checkpoint for fast checkpointing and resuming.
+    distributed = "distributed"
+    # Model state dict, for safe long-term storage in Fast-LLM format.
+    state_dict = "state_dict"
```
Comment (tscholak) [inline fast_llm/engine/config_utils/checkpoint.py:23]: how about we call this `long_term` instead?

File: fast_llm/engine/config_utils/checkpoint.py [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+    def _from_dict(
+        cls,
+        default: dict[str, typing.Any],
+        strict: bool = True,
+        flat: bool = False,
+    ):
+        # TODO v0.2: Remove.
+        if default.get("format", None) == "huggingface":
+            warnings.warn(f"`huggingface` checkpoint format has been renamed to `external`.")
+            default["format"] = CheckpointFormat.external.value
```
Comment (tscholak) [inline fast_llm/engine/config_utils/checkpoint.py:81]: will there be other external checkpoint formats or why this change?

File: fast_llm/engine/config_utils/checkpoint.py [hunk 3]
```diff
@@ -0,0 +1,147 @@
+# TODO: Use packaging.version? (Safer but extra requirement)
```
Comment (tscholak) [inline fast_llm/engine/config_utils/checkpoint.py:1]: not sure where to ask this question:
where will we be configuring the layers-per-step option?

File: fast_llm/engine/config_utils/checkpoint.py [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+logger = logging.getLogger(__name__)
+
+# TODO: Use packaging.version? (Safer but extra requirement)
+CHECKPOINT_VERSION = "0.1"
+KNOWN_CHECKPOINT_VERSIONS = ("0", "0.1")
+
+
+class CheckpointFormat(str, enum.Enum):
+    # Distributed checkpoint for fast checkpointing and resuming.
+    distributed = "distributed"
```
Comment (tscholak) [inline fast_llm/engine/config_utils/checkpoint.py:21]: how about we call this `fast` or `short_term`?

File: fast_llm/engine/multi_stage/fast_llm_model.py [hunk 5]
```diff
@@ -83,28 +85,83 @@ def distributed(self):
...
+            "checkpoint_type": CheckpointFormat.distributed.value,
+            "checkpoint_version": str(CHECKPOINT_VERSION),
+            "fast_llm_config": self._fast_llm_config.to_serialized(),
+            "state_shard_names": list(self._state_shard_names[:num_shards]),
+            "metadata": {} if metadata is None else metadata,
+        }
+
+        # TODO: Simplify branching.
+        if checkpoint_config.format == CheckpointFormat.external:
+            # TODO: Support optimizer?
```
Comment (tscholak) [inline fast_llm/engine/multi_stage/fast_llm_model.py:104]: can you confirm that Fast-LLM will set the `checkpoint_config.optimizer_state` to `False` somewhere (or reject the config at launch) before we end up here? Would be frustrating to crash thousands of steps into training because of this.

File: fast_llm/engine/multi_stage/fast_llm_model.py [hunk 6]
```diff
@@ -83,28 +85,83 @@ def distributed(self):
...
+            ):
+                shard_split = self._state_shard[i].split(self._stage_shard_sizes, 0)
+                for stage, shard in zip(self._stages_on_device.values(), shard_split):
+                    for name, tensor in stage._export_shard(shard, dtype=checkpoint_config.data_type):  # noqa
+                        assert name not in fast_llm_state_dict
+                        fast_llm_state_dict[(name, shard_name)] = tensor
+                        for exported_name, exported_tensor in converter.convert_state_dict(
+                            fast_llm_state_dict, True
+                        ).items():
+                            context.add_tensor(exported_name, exported_tensor)
```
Comment (tscholak) [inline fast_llm/engine/multi_stage/fast_llm_model.py:155]: so "external" is just state-dict saving with ad-hoc conversion?

in that case, do we need "external" as its own format?

right now:

1. `distributed`: fast internal checkpoint format, only for short-term use (i.e. within the same run and with the same (minor?) version of Fast-LLM)
2. `state_dict`: long-term internal checkpoint format, safe to use with different versions of Fast-LLM
3. `external`: like `state_dict`, but potentially lossy because optimizer states cannot be saved and tensors may be converted in a numerically lossy way.

did I capture that correctly?

File: fast_llm/engine/training/config.py [hunk 7]
```diff
@@ -17,48 +25,85 @@
...
-            default=0,
-            desc=f"Offset for the first {offset_desc or desc}.",
-            hint=FieldHint.feature,
-            valid=check_field(Assert.geq, 0),
+@config_class()
+class IntervalConfig(Config):
+    # Intervals are a common pattern, so we standardize them with this base class.
+    interval: int | None = Field(
+        default=None,
+        desc="The number of training iterations between each interval. Setting to None will disable.",
```
Comment (tscholak) [inline fast_llm/engine/training/config.py:33]: maybe the wrong place to do this, but can we make it so that there is at least a warning that checkpoint saving is disabled? I don't think people will appreciate training for hours (or days) only to find out that they forgot to set a saving interval ;)

File: fast_llm/engine/training/config.py [hunk 8]
```diff
@@ -17,48 +25,85 @@
...
+            environment.update(self.environment)
+            subprocess.Popen(self.script, env=environment)
+
+
+@config_class()
+class WandbAlertConfig(IntervalConfig):
+    interval = FieldUpdate(
+        desc="The number of training iterations between each Wandb status post (alert)."
+        " Setting to None will disable iteration-based wandb alerts."
+        " Must be a sub-interval of the logging interval."
```
Comment (tscholak) [inline fast_llm/engine/training/config.py:102]: you mean that wandb posting can only happen at logging times? That would be super-interval, no? can you clarify what you mean here? I'm confused.

File: fast_llm/engine/training/config.py [hunk 9]
```diff
@@ -90,71 +141,109 @@ class WandbConfig(Config):
...
-        )
+        return (self.get_count(training_iterations) + extra_validations) * self.iterations if self.enabled() else 0
 
 
 @config_class()
-class CheckpointConfig(get_interval_config_class("checkpoint")):
+class CheckpointBaseConfig(IntervalConfig):
+    _abstract = True
+    save_name: typing.ClassVar[str] = "save"
+    directory_name: typing.ClassVar[str] = "save"
```
Comment (tscholak) [inline fast_llm/engine/training/config.py:166]: for what kind of checkpoint will we have `save` as the name of the output directory?

File: fast_llm/engine/training/config.py [hunk 10]
```diff
@@ -90,71 +141,109 @@ class WandbConfig(Config):
...
-    callback: CallbackConfig = Field(
-        default_factory=CallbackConfig,
-        desc="Callback (shell script) to run after export.",
-        hint=FieldHint.core,
+class ExportConfig(CheckpointBaseConfig, CheckpointConfigBase, CheckpointStateConfigBase, CheckpointSaveConfigBase):
+    _abstract = False
+    save_name: typing.ClassVar[str] = "export"
+    directory_name = "export"
+    interval = FieldUpdate(
+        desc="The number of training iterations between each export." " Setting to None will disable exports."
```
Comment (tscholak) [inline fast_llm/engine/training/config.py:228]: Can this interval be incompatible with the checkpointing interval (i.e., not an exact multiple or divisor of the checkpointing interval)?

File: fast_llm/engine/training/config.py [hunk 11]
```diff
@@ -90,71 +141,109 @@ class WandbConfig(Config):
...
 
 
 @config_class()
-class ShutdownConfig(get_interval_config_class("automated shutdown")):
-    pass
+class ShutdownConfig(IntervalConfig):
+    interval = FieldUpdate(
+        desc="The number of training iterations between each automated shutdown."
+        " Setting to None will disable automated shutdowns."
+        " Must be a sub-interval of the checkpoint interval."
```
Comment (tscholak) [inline fast_llm/engine/training/config.py:242]: what's an automated shutdown? I am not familiar with this feature? is this new?

**Review by tscholak — COMMENTED** [review]
Great changes, thanks @jlamypoirier.
I have a few suggestions and questions for clarification. Can you please answer them?

File: fast_llm/engine/config_utils/checkpoint.py (hunk 3 above)
Comment (jlamypoirier) [reply fast_llm/engine/config_utils/checkpoint.py:1]: That's a hack for the single process converter, it doesn't apply here because we already have the whole model loaded. I'm not entirely sure about memory usage but it should be OK because we reconstruct one layer at the time.

File: fast_llm/engine/config_utils/checkpoint.py (hunk 4 above)
Comment (jlamypoirier) [reply fast_llm/engine/config_utils/checkpoint.py:21]: Also an option, let's discuss it later.

File: fast_llm/engine/multi_stage/fast_llm_model.py (hunk 5 above)
Comment (jlamypoirier) [reply fast_llm/engine/multi_stage/fast_llm_model.py:104]: Right now there is no safety check, but that would be easy to add. Looking at it, right now the bigger problem is that the optimizer isn't saved by default so the bigger risk is not saving it when we want to...

File: fast_llm/engine/multi_stage/fast_llm_model.py (hunk 6 above)
Comment (jlamypoirier) [reply fast_llm/engine/multi_stage/fast_llm_model.py:155]: Something like that, I'll be clearing things up in the next PR. The plan is to generalize things a bit, something like
```
- AbstractSaver
    - DistributedSaver [hard-coded]
    - AbstractStateDictSaver
        - StateDictSaver [hard-coded]
        - HuggingfaceSaver
            - [model/user-specific]
        - [model/user-specific]
    - [model/user-specific]

Comment (jlamypoirier) [reply fast_llm/engine/multi_stage/fast_llm_model.py:155]: As for the optimizer state, the only reason it's prevented is I haven't implemented it for huggingface, it would probably not be that hard.

File: fast_llm/engine/training/config.py (hunk 8 above)
Comment (jlamypoirier) [reply fast_llm/engine/training/config.py:102]: Yes. By sub-interval I meant posting iterations are a subset of logging ones, but that's probably not the right term.

File: fast_llm/engine/training/config.py (hunk 9 above)
Comment (jlamypoirier) [reply fast_llm/engine/training/config.py:166]: None, that's a placeholder (I could just remove?)

File: fast_llm/engine/training/config.py (hunk 10 above)
Comment (jlamypoirier) [reply fast_llm/engine/training/config.py:228]: Yes it doesn't matter anymore because checkpoints and exports are completely independent.

File: fast_llm/engine/training/config.py (hunk 11 above)
Comment (jlamypoirier) [reply fast_llm/engine/training/config.py:242]: It's always been there but never used.

File: fast_llm/engine/training/config.py (hunk 7 above)
Comment (jlamypoirier) [reply fast_llm/engine/training/config.py:33]: That's an option, but would the warning actually be useful? A runtime warning is unlikely to be seen, and a validation one is only useful if the config is validated before launch.

Comment (tscholak) [reply fast_llm/engine/training/config.py:33]: Hm, I see. Do you have a better idea?

**Review by tscholak — APPROVED** [review]
Thanks @jlamypoirier for answering my comments. Since most changes I suggested will be future work, this can be merged as is.

File: fast_llm/engine/training/config.py (hunk 7 above)
Comment (jlamypoirier) [reply fast_llm/engine/training/config.py:33]: Not really....
